### PR TITLE
Docs/mqtt

### PR DESCRIPTION
**Pull request by michaelboulton**
This should probably be moved into a separate file/page on the website as well

### CODE REVIEW

**Comment by benhowes** [comment]
Can you move it to mqtt.md? @martinalexrolph can you take the lead on reviewing the content?

File: documentation.md [hunk 1]
```diff
@@ -777,3 +777,254 @@ stages:
...
+
+### MQTT connection options
+
+The MQTT library used is the
+[paho-mqtt](https://github.com/eclipse/paho.mqtt.python) Python library, and for
+the most part the arguments for each block are passed directly through to the
+similarly-named methods on the `paho.mqtt.client.Client` class.
+
+The full list of options for the `mqtt` client block are listed below (`host`
+is the only required key, thought you will almost always require some of the
```
Comment (martinalexrolph) [inline documentation.md]: thought -> though

File: documentation.md [hunk 2]
```diff
@@ -777,3 +777,254 @@ stages:
...
+  - `clean_session`: Whether to connect with a clean session or not. `true` or
+    `false`. Defaults to `false`.
+- `connect`: Passed through to `Client.connect`.
+  - `host`: MQTT broker host.
+  - `port`: MQTT broker port. Defaults to 1883 in the paho-mqtt library.
+  - `keepalive`: Keepalive frequency to MQTT broker. Defaults to 60 (seconds) in
+    the paho-mqtt library. Note that some brokers will kick client off after 60
+    seconds by default (eg VerneMQ), so you might need to lower this if you are
+    kicked off frequently.
+  - `timeout`: How long to try and connect to the MQTT broker before giving up.
```
Comment (martinalexrolph) [inline documentation.md]: 'How many seconds to...'

Comment (martinalexrolph) [reply documentation.md]: Or 'How long in seconds to...'

File: documentation.md [hunk 3]
```diff
@@ -777,3 +777,254 @@ stages:
...
+    This is not passed through to paho-mqtt, it is implemented in Tavern.
+    Defaults to 1 (second).
+- `tls`: Controls TLS connections - at the time of writing this only has one key
+  that can be used to enable TLS.
```
Comment (martinalexrolph) [inline documentation.md]: Drop the bit after the '-', it's clear that there's only one key and there's no point drawing attention to the limitations here IMO.

File: documentation.md [hunk 4]
```diff
@@ -777,3 +777,254 @@ stages:
...
+  - `enable`: Enable TLS connection with broker. If enabled, Tavern essentially
+    just calls `Client.tls_enable()`.
```
Comment (martinalexrolph) [inline documentation.md]: In general, don't downplay how Tavern's capabilities. Here, for example, you can drop the 'essentially just' which makes it sound like you think it's not very good.

File: documentation.md [hunk 5]
```diff
@@ -777,3 +777,254 @@ stages:
...
+If other messages on the same topic but with a different payload arrive in the
+meantime, they are ignored and a warning will be logged.
+
+The keys which can be used:
+
+- `topic`: The MQTT topic to subcribe to
+- `payload` OR `json`: A plain text payload or a YAML object that will be
+  serialized into JSON that must match the payload of a message published to
+  `topic`.
+- `timeout`: How long to wait for the message to arrive. Defaults to 3
```
Comment (martinalexrolph) [inline documentation.md]: Again, 'How many seconds to...' or 'How long in seconds to...'

File: index.md [hunk 6]
```diff
@@ -1,6 +1,6 @@
-# Restful API testing
+# Easier API testing
 
-Tavern is a pytest plugin, command-line tool and Python 3 library for automated testing of RESTful APIs, with a simple, concise and flexible YAML-based syntax. It's very simple to get started, and highly customisable for complex tests.
+Tavern is a pytest plugin, command-line tool and Python 3 library for automated testing of APIs, with a simple, concise and flexible YAML-based syntax. It's very simple to get started, and highly customisable for complex tests. At the time of writing, Tavern supports testing RESTful APIs as well as MQTT based APIs.
```
Comment (martinalexrolph) [inline index.md]: I suggest dropping 'At the time of writing'.

File: index.md [hunk 7]
```diff
@@ -51,9 +51,9 @@ $ tavern-ci --stdout minimal_test.tavern.yaml
 
 ## Why not Postman, Insomnia or pyresttest etc?
 
-Tavern is a focused tool which does one thing well: automated testing of RESTful APIs.
+Tavern is a focused tool which does one thing well: automated testing of APIs.
 
-**Postman** and **Insomnia** are excellent tools which cover a wide range of use-cases, and indeed we use Tavern alongside Postman. However, specifically with regards to automated testing, Tavern has several advantages over Postman:
+**Postman** and **Insomnia** are excellent tools which cover a wide range of use-cases for RESTful APIs, and indeed we use Tavern alongside Postman. However, specifically with regards to automated testing, Tavern has several advantages over Postman:
 - A full-featured Python environment for writing custom validation functions
```
Comment (martinalexrolph) [inline index.md]: I suggest mentioning MQTT as the second point for why Tavern is better than Postman. It's a big deal, don't shy away from it!

File: index.md [hunk 8]
```diff
@@ -66,6 +66,8 @@ Tavern does not do many of the things Postman and Insomnia do. For example, Tave
 - Validation function are more flexible and easier to use
 - Better explanations of why a test failed
 
+With the release of version `0.4.0`, Tavern now also supports testing MQTT messages alongside RESTful APIs, a feature missing from the previously discussed tools.
```
Comment (martinalexrolph) [inline index.md]: Drop this paragraph in favour of the bullet point above.

File: documentation.md (hunk 3 above)
Comment (michaelboulton) [reply documentation.md]: We do want to say that in future it should support all the paho options for TLS like passing custom certificates etc, not sure what the best way to say this is... https://github.com/eclipse/paho.mqtt.python/blob/e9914a759f9f5b8081d59fd65edfd18d229a399e/src/paho/mqtt/client.py#L636

File: documentation.md [hunk 9]
```diff
@@ -851,3 +851,252 @@ stages:
...
+  - `host`: MQTT broker host.
+  - `port`: MQTT broker port. Defaults to 1883 in the paho-mqtt library.
+  - `keepalive`: Keepalive frequency to MQTT broker. Defaults to 60 (seconds) in
+    the paho-mqtt library. Note that some brokers will kick client off after 60
+    seconds by default (eg VerneMQ), so you might need to lower this if you are
+    kicked off frequently.
+  - `timeout`: How many seconds to try and connect to the MQTT broker before giving up.
+    This is not passed through to paho-mqtt, it is implemented in Tavern.
+    Defaults to 1.
+- `tls`: Controls TLS connection - in future this will support .
```
Comment (martinalexrolph) [inline documentation.md]: I think you forgot to finish a sentence...

File: index.md [hunk 10]
```diff
@@ -51,16 +51,17 @@ $ tavern-ci --stdout minimal_test.tavern.yaml
...
+- A full-featured Python environment for writing easily reusable custom validation functions
+- Testing of MQTT based systems in tandem with RESTful APIS.
 - Seamless integration with pytest to keep all your tests in one place
 - A simpler, less verbose and clearer testing language
 
 Tavern does not do many of the things Postman and Insomnia do. For example, Tavern does not have a GUI nor does it do API monitoring or mock servers. On the other hand, Tavern is free and open-source and is a more powerful tool for developers to automate tests.
 
-**pyresttest** is similar to Tavern but is no longer actively developed. Tavern also has several advantages over PyRestTest which overall add up to a better developer experience:
+**pyresttest** is a similar to Tavern for testing RESTful APIs, but is no longer actively developed. On top of MQTT testing, Tavern has several other advantages over PyRestTest which overall add up to a better developer experience:
```
Comment (martinalexrolph) [inline index.md]: either 'is a similar tool to...' or 'is similar to...'

File: documentation.md (hunk 9 above)
Comment (michaelboulton) [reply documentation.md]: I put this in the original review response but now it doesn't show up:

We do want to say that in future it should support all the paho options for TLS like passing custom certificates etc, not sure what the best way to say this is... https://github.com/eclipse/paho.mqtt.python/blob/e9914a759f9f5b8081d59fd65edfd18d229a399e/src/paho/mqtt/client.py#L636

Comment (martinalexrolph) [reply documentation.md]: Perhaps:
```
`tls`: Enable or disable the TLS connection. Support for all `paho-mqtt` options, including custom certificates, is planned and will be implemented as new options under this key.
```

**Pull request merged**